### PR TITLE
implements dump which dumps a function as a binary chunk

### PR DESCRIPTION
**Pull request by katdrobnjakovic**
Adds support for [`Dump`](http://www.lua.org/manual/5.3/manual.html#lua_dump)

The corresponding C Lua code can be found in ldump.c. 

cc @camilo

### CODE REVIEW

File: dump.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	d.writeByte(byte(p.maxStackSize))
+	d.writeCode(p)
+	d.writeConstants(p)
+	d.writeUpvalues(p)
+	d.writeDebug(p)
+
+}
+
+func (d *dumpState) dumpHeader() {
+	d.err = binary.Write(d.out, endianness(), header)
```
Comment (fbogsany) [inline dump.go]: Can replace the call to `endianness()` here with `d.order`.

File: dump.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+	d.writeInt(p.lineDefined)
+	d.writeInt(p.lastLineDefined)
+	d.writeByte(byte(p.parameterCount))
+	d.writeBool(p.isVarArg)
+	d.writeByte(byte(p.maxStackSize))
+	d.writeCode(p)
+	d.writeConstants(p)
+	d.writeUpvalues(p)
+	d.writeDebug(p)
+
```
Comment (fbogsany) [inline dump.go]: Unnecessary whitespace. You probably need to run `gofmt`.

File: dump.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+		d.write(uint32(len(s)))
+	default:
+		panic(fmt.Sprintf("unsupported pointer size (%d)"))
+	}
+}
+
+func (d *dumpState) writeDebug(p *prototype) {
+	var n int
+	d.writeString(p.source)
+	d.writeInt(n)
```
Comment (fbogsany) [inline dump.go]: This will output `0` (as will all the other `d.writeInt(n)` calls below). This one should be `d.write(len(p.lineInfo))`.

File: dump.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	d.write(p.lineInfo)
+	d.writeInt(n)
+
+	for _, lv := range p.localVariables {
```
Comment (fbogsany) [inline dump.go]: The `d.writeInt` call above belongs with this line - it should be `d.writeInt(len(p.localVariables))`.

File: dump.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+		d.writeString(lv.name)
+		d.writePC(lv.startPC)
+		d.writePC(lv.endPC)
+	}
+
+	d.writeInt(n)
+
+	for _, uv := range p.upValues {
```
Comment (fbogsany) [inline dump.go]: Likewise here - should be `d.writeInt(len(p.upValues))`

File: lua.go [hunk 6]
```diff
@@ -437,6 +437,19 @@ func (l *State) Load(r io.Reader, chunkName string, mode string) error {
...
+// the top of the stack and produces a binary chunk that, if loaded again,
+// results in a function equivalent to the one dumped.
+//
+// http://www.lua.org/manual/5.3/manual.html#lua_dump
+func (l *State) Dump(w io.Writer) error {
+	l.checkElementCount(1)
+	if f, ok := l.stack[l.top-1].(*luaClosure); ok {
+		return l.dump(f.prototype, w)
+	}
+	return errors.New("todo")
```
Comment (fbogsany) [inline lua.go]: Thoughts on what `"todo"` should be here?

File: dump_test.go [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+	}
+	if closure == nil {
+		t.Error("closure was nil")
+	}
+	undumpedPrototype := closure.prototype
+	if undumpedPrototype == nil {
+		t.Fatal("prototype was nil")
+	}
+
+	if reflect.DeepEqual(p, undumpedPrototype) {
```
Comment (fbogsany) [inline dump_test.go]: This check is wrong - it should be an error if `!reflect.DeepEqual(p, undumpedPrototype)`. Unfortunately, this is passing because the earlier code is wrong. We're dumping an empty `prototype`, then loading a prototype from a file generated by `luac`, then comparing them (incorrectly).

We should:
- load from `lua-tests/checktable.lua` (no need to run the `luac` compiler)
- call `l.Dump(out)` to dump to the `out` byte buffer
- call `l.undump(out, "test")` to load from the dumped prototype
- compare with `reflect.DeepEqual`.

File: dump_test.go [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+	if err != nil {
+		t.Error("error reading file", err)
+	}
+
+	actualContent, err := ioutil.ReadAll(out)
+	if err != nil {
+		t.Error("error reading out bugger", err)
+	}
+
+	if bytes.Equal(expectedContent, actualContent) {
```
Comment (fbogsany) [inline dump_test.go]: This check is also wrong. It should be `if !bytes.Equal(expectedContent, actualContent) {`.

File: dump_test.go [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+		t.Fatal("prototype was nil")
+	}
+
+	out := new(bytes.Buffer)
+	err = l.dump(p, out)
+	if err != nil {
+		t.Error("unexpected error", err, "with testing dump")
+	}
+
+	expectedContent, err := ioutil.ReadFile(source)
```
Comment (fbogsany) [inline dump_test.go]: This should read `binary` rather than `source`.

**Comment by katdrobnjakovic** [comment]
updated with pr comments. ready for 👀

File: undump.go [hunk 10]
```diff
@@ -138,6 +128,32 @@ func (state *loadState) readDebug(p *prototype) (source string, lineInfo []int32
...
+func (state *loadState) readLineInfo() (lineInfo []int32, err error) {
+	var n int32
+	if n, err = state.readInt(); err != nil || n == 0 {
+		return
+	}
+	lineInfo = make([]int32, n)
+	if err = state.read(lineInfo); err != nil {
+		return
+	}
+	return
```
Comment (fbogsany) [inline undump.go]: You can just:
```go
lineInfo = make([]int32, n)
err = state.read(lineInfo)
return
```

File: lua.go [hunk 11]
```diff
@@ -437,6 +437,19 @@ func (l *State) Load(r io.Reader, chunkName string, mode string) error {
...
+// the top of the stack and produces a binary chunk that, if loaded again,
+// results in a function equivalent to the one dumped.
+//
+// http://www.lua.org/manual/5.3/manual.html#lua_dump
+func (l *State) Dump(w io.Writer) error {
+	l.checkElementCount(1)
+	if f, ok := l.stack[l.top-1].(*luaClosure); ok {
+		return l.dump(f.prototype, w)
+	}
+	return errors.New("error in dumping the function as a binary chunk")
```
Comment (fbogsany) [inline lua.go]: IMO totally reasonable to `panic("closure expected")` in this case, rather than `return errors.New...`

File: dump.go [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+	d.writeString(p.source)
+	d.writeInt(len(p.lineInfo))
+	d.write(p.lineInfo)
+	d.writeInt(len(p.localVariables))
+
+	for _, lv := range p.localVariables {
+		d.writeString(lv.name)
+		d.writePC(lv.startPC)
+		d.writePC(lv.endPC)
+	}
```
Comment (fbogsany) [inline dump.go]: Could break this out into `writeLocalVariables` for consistency with undump's `readLocalVariables`.

File: dump.go [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+		default:
+			d.l.assert(false)
+		}
+	}
+
+	d.writeInt(len(p.prototypes))
+
+	for _, o := range p.prototypes {
+		d.dumpFunction(&o)
+	}
```
Comment (fbogsany) [inline dump.go]: Could break this out into `writePrototypes` for consistency with `readPrototypes` in `undump.go`.

File: undump.go [hunk 14]
```diff
@@ -176,7 +193,11 @@ func (state *loadState) readConstants() (constants []value, prototypes []prototy
 			return
 		}
 	}
-	if n, err = state.readInt(); err != nil {
+	return
+}
+func (state *loadState) readPrototypes() (prototypes []prototype, err error) {
```
Comment (fbogsany) [inline undump.go]: Add a blank line before this `func`.

File: dump.go (hunk 13 above)
Comment (katdrobnjakovic) [inline dump.go]: done :)

File: dump.go (hunk 12 above)
Comment (katdrobnjakovic) [inline dump.go]: done

**Comment by katdrobnjakovic** [comment]
updated with pr comments
